### PR TITLE
Add module asset registry

### PR DESCRIPTION
**Pull request by zjb0807**
Closes: #1574 

TODO:
- [x] benchmarking
- [x] EvmAddress support `DexShare::LiquidCroadloan` and `DexShare::ForeignAsset`
- [x] Intergration tests

### CODE REVIEW

File: modules/asset-registry/src/lib.rs [hunk 1]
```diff
@@ -51,17 +64,74 @@ pub mod module {
...
+	#[pallet::getter(fn multi_locations)]
+	pub type MultiLocations<T: Config> = StorageMap<_, Twox64Concat, ForeignAssetId, MultiLocation, OptionQuery>;
+
+	/// The storages for AssetMetadatas.
+	///
+	/// AssetMetadatas: map MultiLocation => Option<AssetMetadata>
+	#[pallet::storage]
+	#[pallet::getter(fn asset_metadatas)]
+	pub type AssetMetadatas<T: Config> =
+		StorageMap<_, Twox64Concat, MultiLocation, AssetMetadata<BalanceOf<T>>, OptionQuery>;
```
Comment (xlc) [inline modules/asset-registry/src/lib.rs:133]: use `ForeignAssetId` as key can reduce migration required when upgrading MultiLocation in later xcm version

**Comment by zjb0807** [comment]
Try to integrate `statemine` to test cross-chain transfer in another PR

**Review by shaunxw — COMMENTED** [review]
Note `v1::MultiLocation` is used in storages, we would need to do migration if upgrade to `v2` in the future.

**Comment by xlc** [comment]
We deliberately use v1:MultiLocation. v2 is just reexport of v1.
We want to avoid the VersionedMultiLocation because it will not be unique.

File: modules/asset-registry/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+
+				let mut vec = Vec::new();
+				vec.extend_from_slice(&b"LP "[..]);
+				vec.extend_from_slice(&name_0);
+				vec.extend_from_slice(&b" - ".to_vec());
+				vec.extend_from_slice(&name_1);
+				Some(vec)
+			}
+			CurrencyId::Erc20(address) => Pallet::<T>::get_erc20_mapping(address).map(|v| v.name),
+			CurrencyId::StableAssetPoolToken(_) => None,
```
Comment (xlc) [inline modules/asset-registry/src/lib.rs:349]: we should have an issue and TODO for this

**Review by xlc — APPROVED** [review]

**Review by shaunxw — DISMISSED** [review]

**Review by shaunxw — APPROVED** [review]

**Pull request merged**